### DISCUSSION TITLE
Interaction of Cloudflare worker and RDS backend

### DISCUSSION THREAD

**Discussion by RitikJaiswal75** [Real-Dev-Squad/discord-slash-commands#9] [General]
# Sending and receiving data between Cloudflare worker and RDS backend.

Here we are looking for a way to send and receive data between the RDS backend and Cloudflare worker. We need the means to verify the calls are coming only from the specified Cloudflare worker. 
The method we chose was using RSA Private and Public keys and generating a JWT for the requests.

![server to server drawio](https://user-images.githubusercontent.com/57758447/213845674-692ad14a-989e-4aa4-9c32-dca42bd782dc.png)

**Comment by rohan09-raj** [comment]
## Design Frontend
https://www.figma.com/file/5Tk9bgA7DXchDtCdrrPSwD/Main-Site?type=design&node-id=644-14&mode=design&t=OkhGdZVK2YRBDCsI-0

## Implementation
Implemented the working code for the successful interaction between `Real Dev Squad Backend` and `Cloudflare Worker`  

## JWT Interaction Details
### Cloudflare Workers
- Package Used -> https://www.npmjs.com/package/@tsndr/cloudflare-worker-jwt 
- Method -> `sign`
- Payload -> `{name: "Cloudflare Worker"}`
- Key Type -> `Private Key` (as it is generated)
- Algorithm -> `RS256`
- Expiry -> `2 seconds`
- Code Sample ->
```
const name = { name: "Cloudflare Worker" };
const jwtToken = await jwt.sign(
      { 
        name: "Cloudflare Worker", 
        exp: Math.floor(Date.now() / 1000) + 2 
      }, 
      env.PRIVATE_KEY, 
      { 
        algorithm: 'RS256' 
      }
);
```

### Backend
- Package Used -> https://www.npmjs.com/package/jsonwebtoken
- Method -> `verify`
- Key Type -> `Public Key` (Append string with \n and +)
- Algorithm -> `RS256`
- Code Sample ->
```
const headers = req.headers.authorization.split(" ");
const token = headers[1];
const data = jwt.verify(token, config.get("botToken.publicKey"), { algorithms: ["RS256"] });
```
